### PR TITLE
Add acceptance tests for k8s permissions

### PR DESCRIPTION
**Pull request by nimrod-up9**
Test k8s permissions. After this PR if a developer adds a function that requires additional permissions (new verb on old noun or new noun), then the acceptance tests will fail and we will know to update the permissions lists.

All tests are run by default with the permissions that are mandatory for `mizu tap`. These are the permissions that `mizu check --pre-tap` checks. One test tests is run in restricted mode with its permissions. One test runs with the additional permissions that are required for resolving and checks that it works.

Note that there are only positive tests. This PR does not add tests to check what happens if there are missing permissions.

This PR does not check all permission YAMLs. It does not check `permissions-ns-ip-resolution-optional.yaml` because it is very similar to `permissions-all-namespaces-ip-resolution-optional.yaml` and we can save the time it takes to run the test (~2m). The PR does not check `*debug-optional.yaml` files.

### CODE REVIEW

File: acceptanceTests/tap_test.go [hunk 1]
```diff
@@ -632,3 +632,34 @@ func TestTapDumpLogs(t *testing.T) {
...
+
+func TestRestrictedMode(t *testing.T) {
+	t.Log("creating permissions for restricted user")
+	if err := applyKubeFilesForTest(
+		t,
+		"minikube",
+		"mizu-tests",
+		"../examples/roles/permissions-ns-tap.yaml",
+		"../examples/roles/permissions-ns-ip-resolution-optional.yaml",
+		"../examples/roles/permissions-ns-debug-optional.yaml",
```
Comment (nimrod-up9) [inline acceptanceTests/tap_test.go:644]: remove optional permissions

File: cli/cmd/permissionFiles/permissions-ns-tap.yaml [hunk 2]
```diff
@@ -20,15 +19,17 @@ rules:
 - apiGroups: [""]
   resources: ["configmaps"]
   verbs: ["create", "delete"]
+- apiGroups: [""]
+  resources: ["pods/log"]
+  verbs: ["get"]
```
Comment (nimrod-up9) [inline cli/cmd/permissionFiles/permissions-ns-tap.yaml:24]: Missing permissions.
Caught by the acceptance tests :)

File: cli/cmd/permissionFiles/permissions-all-namespaces-tap.yaml [hunk 3]
```diff
@@ -22,14 +22,17 @@ rules:
 - apiGroups: [""]
   resources: ["configmaps"]
   verbs: ["create"]
+- apiGroups: [""]
+  resources: ["pods/log"]
+  verbs: ["get"]
```
Comment (nimrod-up9) [inline cli/cmd/permissionFiles/permissions-all-namespaces-tap.yaml:27]: Missing permissions.
Caught by the acceptance tests :)

File: acceptanceTests/testsUtils.go [hunk 4]
```diff
@@ -103,14 +111,96 @@ func getServiceExternalIp(ctx context.Context, namespace string, service string)
...
+	}
+
+	if err := setKubeCurrentContext(newContextName); err != nil {
+		return err
+	}
+
+	t.Cleanup(func() {
+		err := setKubeCurrentContext(prevKubeContextName)
+		if err != nil {
+			t.Errorf("failed to set Kubernetes context to %s, err: %v", prevKubeContextName, err)
```
[the remark this answers is not in the document]
Comment (nimrod-up9) [reply acceptanceTests/testsUtils.go:127]: done

File: .github/workflows/acceptance_tests.yml [hunk 5]
```diff
@@ -21,20 +22,57 @@ jobs:
...
         run: source ./acceptanceTests/setup.sh
 
+      - name: Create k8s users
+        env:
+          USERNAME_UNRESTRICTED: admin
+          USERNAME_RESTRICTED: user1
+        run: |
+          ./acceptanceTests/create_user.sh "${USERNAME_UNRESTRICTED}"
+          ./acceptanceTests/create_user.sh "${USERNAME_RESTRICTED}"
+          sed s/user1/admin/ examples/roles/permissions-all-namespaces-tap.yaml | kubectl apply -f -
```
[the remark this answers is not in the document]
Comment (nimrod-up9) [reply .github/workflows/acceptance_tests.yml]: done

File: .github/workflows/acceptance_tests.yml [hunk 6]
```diff
@@ -21,20 +22,57 @@ jobs:
...
+          chmod 755 ./cli/bin/mizu_ci
+          ls -l /home/runner/work/mizu/mizu/cli/bin/mizu_ci
+
+
       - name: Setup acceptance test
         run: source ./acceptanceTests/setup.sh
 
+      - name: Create k8s users
+        env:
+          USERNAME_UNRESTRICTED: admin
```
[the remark this answers is not in the document]
Comment (nimrod-up9) [reply .github/workflows/acceptance_tests.yml]: done

File: acceptanceTests/create_user.sh [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+	 sudo apt-get install openssl
+fi
+
+echo "Creating certificate for user \"${NEW_USERNAME}\""
+mkdir -p ${CERT_DIR}
+echo "Generating key \"${KEY_FILE}\""
+openssl genrsa -out "${KEY_FILE}" 2048
+echo "Generating crt \"${CRT_FILE}\""
+openssl req -new -key "${KEY_FILE}" -out "${CRT_FILE}" -subj "/CN=${NEW_USERNAME}/O=group1"
+openssl x509 -req -in "${CRT_FILE}" -CA "${MINIKUBE_CRT_FILE}" -CAkey "${MINIKUBE_KEY_FILE}" -CAcreateserial -out "${CRT_FILE}" -days 500
```
Comment (RoyUP9) [inline acceptanceTests/create_user.sh:32]: is there a reason for the 500 days?

File: acceptanceTests/extensions_test.go [hunk 8]
```diff
@@ -49,7 +49,12 @@ func TestRedis(t *testing.T) {
 
 	ctx := context.Background()
 
-	redisExternalIp, err := getServiceExternalIp(ctx, defaultNamespaceName, "redis")
+	kubernetesProvider, err := newKubernetesProvider()
+	if err != nil {
+		t.Errorf("failed to create k8s provider, err %v", err)
```
Comment (RoyUP9) [inline acceptanceTests/extensions_test.go:54]: add return after

File: acceptanceTests/extensions_test.go [hunk 9]
```diff
@@ -141,7 +146,12 @@ func TestAmqp(t *testing.T) {
 
 	ctx := context.Background()
 
-	rabbitmqExternalIp, err := getServiceExternalIp(ctx, defaultNamespaceName, "rabbitmq")
+	kubernetesProvider, err := newKubernetesProvider()
+	if err != nil {
+		t.Errorf("failed to create k8s provider, err %v", err)
```
Comment (RoyUP9) [inline acceptanceTests/extensions_test.go:151]: add return after

File: acceptanceTests/tap_test.go [hunk 10]
```diff
@@ -632,3 +638,26 @@ func TestTapDumpLogs(t *testing.T) {
...
+
+	t.Log("creating permissions for restricted user")
+	if err := applyKubeFilesForTest(
+		t,
+		"minikube",
+		namespace,
+		"../cli/cmd/permissionFiles/permissions-ns-tap.yaml",
+		"../cli/cmd/permissionFiles/permissions-ns-ip-resolution-optional.yaml",
+	); err != nil {
+		t.Errorf("failed to create k8s permissions, %v", err)
```
Comment (RoyUP9) [inline acceptanceTests/tap_test.go:653]: add return after

File: acceptanceTests/tap_test.go [hunk 11]
```diff
@@ -632,3 +638,26 @@ func TestTapDumpLogs(t *testing.T) {
...
+	}
+
+	t.Log("switching k8s context to user")
+	if err := switchKubeContextForTest(t, "user-with-restricted-access"); err != nil {
+		t.Errorf("failed to switch k8s context, %v", err)
```
Comment (RoyUP9) [inline acceptanceTests/tap_test.go:658]: add return after

File: acceptanceTests/tap_test.go [hunk 12]
```diff
@@ -632,3 +638,26 @@ func TestTapDumpLogs(t *testing.T) {
...
+		"minikube",
+		namespace,
+		"../cli/cmd/permissionFiles/permissions-ns-tap.yaml",
+		"../cli/cmd/permissionFiles/permissions-ns-ip-resolution-optional.yaml",
+	); err != nil {
+		t.Errorf("failed to create k8s permissions, %v", err)
+	}
+
+	t.Log("switching k8s context to user")
+	if err := switchKubeContextForTest(t, "user-with-restricted-access"); err != nil {
```
Comment (RoyUP9) [inline acceptanceTests/tap_test.go:657]: don't we need to switch back to the no restricted access user at the cleanup of the test

File: acceptanceTests/testsUtils.go [hunk 13]
```diff
@@ -103,6 +111,90 @@ func getServiceExternalIp(ctx context.Context, namespace string, service string)
...
+	if err != nil {
+		return err
+	}
+
+	if err := setKubeCurrentContext(newContextName); err != nil {
+		return err
+	}
+
+	t.Cleanup(func() {
+		err := setKubeCurrentContext(prevKubeContextName)
```
Comment (RoyUP9) [inline acceptanceTests/testsUtils.go:125]: can be changed to single line
if ; err != nil

File: acceptanceTests/testsUtils.go [hunk 14]
```diff
@@ -103,6 +111,90 @@ func getServiceExternalIp(ctx context.Context, namespace string, service string)
...
+		if err != nil {
+			t.Errorf("failed to set Kubernetes context to %s, err: %v", prevKubeContextName, err)
+			t.Errorf("cleanup failed, subsequent tests may be affected")
```
Comment (RoyUP9) [inline acceptanceTests/testsUtils.go:128]: maybe add this log to the one above

File: acceptanceTests/create_user.sh (hunk 7 above)
Comment (nimrod-up9) [reply acceptanceTests/create_user.sh:32]: No, it was in the example.

File: acceptanceTests/extensions_test.go (hunk 8 above)
Comment (gadotroee) [inline acceptanceTests/extensions_test.go:54]:
```suggestion
		t.Errorf("failed to create k8s provider, err %v", err)
		return
```

File: acceptanceTests/extensions_test.go (hunk 9 above)
Comment (gadotroee) [inline acceptanceTests/extensions_test.go:151]:
```suggestion
		t.Errorf("failed to create k8s provider, err %v", err)
		return
```

File: acceptanceTests/tap_test.go (hunk 10 above)
Comment (gadotroee) [inline acceptanceTests/tap_test.go:653]:
```suggestion
		t.Errorf("failed to create k8s permissions, %v", err)
		return
```

File: acceptanceTests/tap_test.go (hunk 11 above)
Comment (gadotroee) [inline acceptanceTests/tap_test.go:658]:
```suggestion
		t.Errorf("failed to switch k8s context, %v", err)
		return
```

File: acceptanceTests/testsUtils.go [hunk 15]
```diff
@@ -103,6 +111,90 @@ func getServiceExternalIp(ctx context.Context, namespace string, service string)
...
+		}
+	})
+
+	return nil
```
Comment (gadotroee) [inline acceptanceTests/testsUtils.go:132]:
```suggestion
		if err := setKubeCurrentContext(prevKubeContextName); err != nil {
			t.Errorf("failed to set Kubernetes context to %s, err: %v", prevKubeContextName, err)
			t.Errorf("cleanup failed, subsequent tests may be affected")
		}
	})

	return nil
```

File: acceptanceTests/testsUtils.go (hunk 14 above)
Comment (nimrod-up9) [reply acceptanceTests/testsUtils.go:127]: I prefer it like this. I think of it as a separate message.

File: acceptanceTests/testsUtils.go (hunk 13 above)
Comment (nimrod-up9) [reply acceptanceTests/testsUtils.go]: done (by Roee)

File: acceptanceTests/tap_test.go (hunk 11 above)
Comment (nimrod-up9) [reply acceptanceTests/tap_test.go:671]: done (by Roee)

File: acceptanceTests/tap_test.go (hunk 10 above)
Comment (nimrod-up9) [reply acceptanceTests/tap_test.go:665]: done (by Roee)

File: acceptanceTests/extensions_test.go (hunk 9 above)
Comment (nimrod-up9) [reply acceptanceTests/extensions_test.go:152]: done (by Roee)

File: acceptanceTests/extensions_test.go (hunk 8 above)
Comment (nimrod-up9) [reply acceptanceTests/extensions_test.go:54]: done (by Roee)

File: acceptanceTests/create_user.sh (hunk 7 above)
Comment (nimrod-up9) [reply acceptanceTests/create_user.sh]: Changed it.

File: acceptanceTests/tap_test.go [hunk 16]
```diff
@@ -14,6 +14,10 @@ import (
 )
 
 func TestTap(t *testing.T) {
+	basicTapTest(t, false)
```
Comment (RoyUP9) [inline acceptanceTests/tap_test.go:17]: why is it false? don't we want to check src and dest in the regular tap?

File: acceptanceTests/tap_test.go [hunk 17]
```diff
@@ -644,3 +649,43 @@ func TestTapDumpLogs(t *testing.T) {
...
+		"minikube",
+		namespace,
+		"../cli/cmd/permissionFiles/permissions-all-namespaces-ip-resolution-optional.yaml",
+	); err != nil {
+		t.Errorf("failed to create k8s permissions, %v", err)
+		return
+	}
+
+	basicTapTest(t, true)
+}
```
Comment (RoyUP9) [inline acceptanceTests/tap_test.go:668]: add space after func

File: acceptanceTests/tap_test.go [hunk 18]
```diff
@@ -644,3 +649,43 @@ func TestTapDumpLogs(t *testing.T) {
 		return
 	}
 }
+
+func TestIpResolving(t *testing.T) {
+	namespace := ""
```
Comment (RoyUP9) [inline acceptanceTests/tap_test.go:654]: maybe create a const that is named all namespaces with empty value

File: acceptanceTests/tap_test.go (hunk 17 above)
Comment (nimrod-up9) [reply acceptanceTests/tap_test.go:668]: done

**Review by gadotroee — APPROVED** [review]

**Pull request merged**